scRNA-seq-advanced/03-dataset_integration.Rmd

```diff
@@ -0,0 +1,515 @@
+  # Update the rowData names by prepending {library_name}, except for `gene_symbol`
+  names(rowData(sce)) <- ifelse(
+     # Is the value is gene_symbol?
+    names(rowData(sce)) == "gene_symbol",      
+    # If yes, keep gene_symbol
+    "gene_symbol",       
+    # If no, update the name by prepending the library name
+    glue::glue("{library_name}-{names(rowData(sce))}")
+  )
+               
+  ###### Ensure cells can be identified ######
+  # Update the colData row names by prepending {library_name}
+  rownames(colData(sce)) <- glue::glue("{library_name}-{rownames(colData(sce))}")
+                 
+  ###### Add in sample-level information as stand-alone column ######
+  colData(sce)$sample <- library_name
+
+  # Return the formatted SCE object
+  return(sce)
+}
+```
+
+To run this function, we'll use the convenient `purrr::imap()` function, which _assumes_ that the second argument to the function we're calling are the list names:
+
+```{r format sces for merge}
+sce_list_formatted <- purrr::imap(
+  sce_list,
+  format_sce_list,
+  shared_genes
+)
+sce_list_formatted
+```
+
+We are now ready to merge the SCEs!
+We'll use the base R function `cbind`, but since the input is a list we'll have to use base R `do.call` to run `cbind` over the list:
+
+
+```{r merge sces}
+merged_sce <- do.call(cbind, sce_list_formatted)
+merged_sce
+```
+We now have a single SCE object that contains all cells from all libraries we'd like to integrate.
+
+## Integration
+
+### First, explore! (needs better title?)
+
+You might have noticed that that there are already `PCA` and `UMAP` dimension reductions in the merged SCE we created.
+These represent the original dimension reductions that were performed on _each individual library_.
+
+We can also calculate PCA and UMAP for the merged SCE as a whole, which we will save with different names `merged_PCA` and `merged_UMAP` to distinguish them for the original.
+When calculating PCs on this merged object, we need to use a slightly different approach that respects the batch differences within. 
+We can compare these UMAP embeddings to those calculated after performing integration.
+
+
+```{r}
+# calculate high-variance genes to use for dimension reduction
+gene_variance <- scran::modelGeneVar(merged_sce,
+                                     # specify the grouping/batch column:
+                                     block = merged_sce$sample)
+hv_list <- scran::getTopHVGs(gene_variance,
+                              n = 2000)
+
+
+# Use batchelor to calculate PCA from a merged group fo SCEs
+multi_pca <- batchelor::multiBatchPCA(merged_sce,
+                                      subset.row = hv_list,
+                                      batch = merged_sce$sample,
+                                      preserve.single = TRUE)
+
+# add PCA results to merged SCE object 
+reducedDim(merged_sce, "merged_PCA") <- multi_pca@listData[[1]]
```

Comment:
I think this can just be `multi_pca[[1]]`. Either way, this is a little strange, so we should say what is going on here. What is the output of `multiBatchPCA`? A list(ish) of matrices but because we gave it a merged SCE, it only has one matrix.
```suggestion
reducedDim(merged_sce, "merged_PCA") <- multi_pca[[1]]
```